func/internal/podevaluator.go

```diff
@@ -320,13 +326,30 @@ func (pcm *podCacheManager) podCacheManager() {
 						klog.Warningf("pod %v/%v is in Failed state. Deleting and recreating", pod.Namespace, pod.Name)
 						go pcm.deleteFailedPod(pod, req.image)
 						deleteCacheEntry = true
-					} else if pod.DeletionTimestamp == nil && net.JoinHostPort(pod.Status.PodIP, defaultWrapperServerPort) == podAndCl.grpcClient.Target() {
-						klog.Infof("reusing the connection to pod %v/%v to evaluate %v", pod.Namespace, pod.Name, req.image)
-						req.grpcClientCh <- &clientConnAndError{grpcClient: podAndCl.grpcClient}
-						go patchPodWithUnixTimeAnnotation(pcm.podManager.kubeClient, podAndCl.pod, pcm.podTTL)
-						break
-					} else {
+					}
+
+					// GRPC client is made to Service IP instead of PoD directly, hence target host to be verified
+					// against Service Cluster IP
+
+					// Service name is Image Label set on Pod manifest
+					serviceName := pod.ObjectMeta.Labels[krmFunctionImageLabel]
+					service := &corev1.Service{}
+					serviceKey := client.ObjectKey{Namespace: pod.Namespace, Name: serviceName}
+					err = pcm.podManager.kubeClient.Get(context.Background(), serviceKey, service)
+					if err != nil {
+						// Service lookup should not fail.
+						// If it does, purge the cached entry and let Pod/Service be created again
 						deleteCacheEntry = true
+					} else {
+
+						if pod.DeletionTimestamp == nil && net.JoinHostPort(service.Spec.ClusterIP, defaultWrapperServerPort) == podAndCl.grpcClient.Target() {
```

Comment:
I think that the clusterIP service's hostname should be used instead of the allocated IP address.

Reply:
Made required changes to use service FQDN url for building GRPC client